temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java

```diff
@@ -139,15 +141,29 @@ public DeterministicRunner getRunner() {
     return runner;
   }
 
-  public WorkflowOutboundCallsInterceptor getWorkflowInterceptor() {
+  public WorkflowOutboundCallsInterceptor getWorkflowOutboundInterceptor() {
     // This is needed for unit tests that create DeterministicRunner directly.
-    return headInterceptor == null ? this : headInterceptor;
+    return headOutboundInterceptor == null ? this : headOutboundInterceptor;
   }
 
-  public void setHeadInterceptor(WorkflowOutboundCallsInterceptor head) {
-    if (headInterceptor == null) {
-      runner.setInterceptorHead(head);
-      this.headInterceptor = head;
+  public WorkflowInboundCallsInterceptor getWorkflowInboundInterceptor() {
+    // This is needed for unit tests that create DeterministicRunner directly.
+    return headInboundInterceptor == null
+        ? new BaseRootWorkflowInboundCallsInterceptor(this)
+        : headInboundInterceptor;
+  }
+
+  public void initHeadOutboundCallsInterceptor(WorkflowOutboundCallsInterceptor head) {
+    if (headOutboundInterceptor == null) {
+      headOutboundInterceptor = head;
+    }
+  }
+
+  public void initHeadInboundCallsInterceptor(WorkflowInboundCallsInterceptor head) {
+    if (headInboundInterceptor == null) {
```

Comment:
Why null check?

Reply:
I actually saved the behavior that already was here. It was a null check inside a setter method which is even worse. Probably it was done to avoid double initialization with different implementation, but I'm not sure. But I agree it looks off. Let me try to rework it to throwing an exception if the interceptor is already initialized and see if anything blows up.

Reply:
changed to a throwing an exception. see the explanation behind.